src/main/kotlin/io/bitvavo/simulation/Main.kt

```diff
@@ -0,0 +1,50 @@
+package io.bitvavo.simulation
+
+import io.bitvavo.simulation.engine.InMemoryMatchingEngine
+import io.bitvavo.simulation.engine.order_book.InMemoryOrderBook
+import io.bitvavo.simulation.exchange.SimulatorExchange
+import io.bitvavo.simulation.exchange.sequence_generator.CounterGenerator
+import io.bitvavo.simulation.extensions.StreamWriter.writeLine
+import io.bitvavo.simulation.formatters.OrderBookFormatter.formatOrderBook
+import io.bitvavo.simulation.formatters.TradeFormatter.formatTrade
+import java.io.*
+
+fun main() {
+    val input = BufferedReader(InputStreamReader(System.`in`))
+    val output = PrintWriter(OutputStreamWriter(System.out))
+    val logs = PrintWriter(FileWriter("logs.info", true))
+
+    try {
+        val simulator = SimulatorExchange(
+            engine = InMemoryMatchingEngine(InMemoryOrderBook()),
+            generator = CounterGenerator()
+        )
+
+        simulator.onTrade = { trades -> trades.forEach { output.writeLine(formatTrade(it)) } }
+        simulator.onFailure = { logs.writeLine(it) }
+        simulator.onSnapshot = { output.writeLine(formatOrderBook(it)) }
+        simulator.use {
+            while (true) {
+                try {
+                    val command = input.readLine() ?: break
+                    simulator.executeCommand(command)
+                }
+                catch (ex: Exception) {
+                    logs.writeLine("Unhandled error occurred: ${ex.message}")
+                }
+                finally {
+                    output.flush()
+                }
+            }
+        }
+    }
+    finally {
+        listOf(input, output, logs).forEach { stream ->
```

Comment:
You can use https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.io/use.html for managing resources in Kotlin

Reply:
To my taste the main procedure has too many try/catches, I would try to remove the top one

Reply:
how? 

I know if use, but when you have multiple resources it does not look good anymore, so I opted in for this list + forEach

as for the try catch - the outer one is there to guarantee a more or less correct disposal of resource, I could not figure out how to use less of them

how do you suggest to remove the outer one? please elaborate.

Reply:
I this this part: 
```try {
          val command = input.readLine() ?: break
          simulator.executeCommand(command)
      }
      catch (ex: Exception) {
          logs.writeLine("Unhandled error occurred: ${ex.message}")
      }
      finally {
          output.flush()
      }```
Can be simplified to: 
```
    val command = input.readLine() ?: break
    try {
        simulator.executeCommand(command)
    } catch (ex: Exception) {
        logs.writeLine("Unhandled error occurred: ${ex.message}")
    }
    output.flush()
```
In my opinion, the top-level try can be removed, you can just execute the channel closure after "white (true)" exits. I  think you are already catching all the exceptions in the try/catch with "val command = input.readLine() ?: break"